File: android/Book-AndroidProgramming/QuizApp/app/src/main/java/kr/ac/ajou/heidi/quizapp/CheatActivity.java
package kr.ac.ajou.heidi.quizapp;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.widget.Button;
import android.widget.TextView;

public class CheatActivity extends AppCompatActivity {

    private static final String EXTRA_ANSWER_IS_TURE = "kr.ac.ajou.heidi.quizapp.answer_is_true";
    private static final String EXTRA_ANSWER_SHOWN = "kr.ac.ajou.heidi.quizapp.answer_shown";
    private boolean mAnswerIsTrue;
    private TextView mAnswerTextView;
    private Button mShowAnswer;
    private TextView mApiLevelTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cheat);

        if (savedInstanceState != null) {
            setAnswerShownResult(savedInstanceState.getBoolean(EXTRA_ANSWER_SHOWN));
        }

        mAnswerIsTrue = getIntent().getBooleanExtra(EXTRA_ANSWER_IS_TURE, false);
        mAnswerTextView = findViewById(R.id.answer_text_view);
        mShowAnswer = findViewById(R.id.show_answer_button);
        mApiLevelTextView = findViewById(R.id.api_level_text_view);
        mApiLevelTextView.setText(String.valueOf("API 레벨" + Build.VERSION.SDK_INT));
        mShowAnswer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mAnswerIsTrue) {
                    mAnswerTextView.setText(R.string.true_button);
                } else {
                    mAnswerTextView.setText(R.string.false_button);
                }
                setAnswerShownResult(true);

                int cx = mShowAnswer.getWidth() / 2;
                int cy = mShowAnswer.getHeight() / 2;
                float radius = mShowAnswer.getWidth();

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    Animator anim = ViewAnimationUtils
                            .createCircularReveal(mShowAnswer, cx, cy, radius, 0);
                    anim.addListener(new AnimatorListenerAdapter() {
                        @Override
                        public void onAnimationEnd(Animator animation) {
                            super.onAnimationEnd(animation);
                            mShowAnswer.setVisibility(View.INVISIBLE);
                        }
                    });
                    anim.start();
                } else {
                    mShowAnswer.setVisibility(View.INVISIBLE);
                }
            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(EXTRA_ANSWER_SHOWN, false);
    }

    public static Intent newIntent(Context packageContext, boolean answerIsTrue) {
        Intent intent = new Intent(packageContext, CheatActivity.class);
        intent.putExtra(EXTRA_ANSWER_IS_TURE, answerIsTrue);
        return intent;
    }

    public static boolean wasAnswerShown(Intent result) {
        return result.getBooleanExtra(EXTRA_ANSWER_SHOWN, false);
    }

    private void setAnswerShownResult(boolean isAnswerShown) {
        Intent data = new Intent();
        data.putExtra(EXTRA_ANSWER_SHOWN, isAnswerShown);
        setResult(RESULT_OK, data);
    }

}

File: android/Book-AndroidProgramming/GeoQuiz/app/src/main/java/kr/ac/ajou/heidi/geoquiz/CheatActivity.kt
package kr.ac.ajou.heidi.geoquiz

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.os.PersistableBundle
import android.util.Log
import kotlinx.android.synthetic.main.activity_cheat.*
import kr.ac.ajou.heidi.geoquiz.CheatActivity.Companion.EXTRA_ANSWER_IS_TURE
import kr.ac.ajou.heidi.geoquiz.CheatActivity.Companion.IS_ANSWER_SHOWN

class CheatActivity : AppCompatActivity() {

    companion object {
       const val EXTRA_ANSWER_IS_TURE = "kr.ac.ajou.heidi.quizapp.answer_is_true"
        const val IS_ANSWER_SHOWN = "kr.ac.ajou.heidi.quizapp.is_answer_shown"
    }

    var isAnswerShown = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cheat)

        savedInstanceState?.let{
            setAnswerShownResult(it.getBoolean(IS_ANSWER_SHOWN))
        }

        val answer = intent.getBooleanExtra("answer", false)

        cheatButton.setOnClickListener {
            if (answer) answerTextView.text = getString(R.string.true_button)
            else answerTextView.text = getString(R.string.false_button)
            isAnswerShown = true
            setAnswerShownResult(isAnswerShown)
        }
    }

    fun setAnswerShownResult(isAnswerShown: Boolean) {
        val data = Intent()
        data.putExtra(IS_ANSWER_SHOWN, isAnswerShown)
        setResult(Activity.RESULT_OK, data)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Log.i("CheatActivity2", "onSaveInstanceState")
        outState?.putBoolean(IS_ANSWER_SHOWN, true)
    }
}

fun newIntent(packageContext: Context, answer: Boolean): Intent {
    val intent = Intent(packageContext, CheatActivity::class.java)
    intent.putExtra(EXTRA_ANSWER_IS_TURE, answer)
    return intent
}

fun wasAnswerShown(result : Intent) : Boolean{
    Log.i("CheatActivity", "${result.getBooleanExtra(EXTRA_ANSWER_IS_TURE, false)}")
    return result.getBooleanExtra(IS_ANSWER_SHOWN, false)
}
